### PR TITLE
Bethe Salpeter equation for molecules

### PR DESCRIPTION
**Pull request by MGraml**
Added Bethe Salpeter on top of GW for the calculation of electronic excitations in molecules.

### CODE REVIEW

File: src/bse_iterative.F [hunk 1]
```diff
@@ -0,0 +1,1788 @@
...
+         WRITE (unit_nr, *) "fac_max_z_space", fac_max_z_space
+         WRITE (unit_nr, *) "num_new_t", num_new_t
+         WRITE (unit_nr, *) "eps_res", eps_res
+         WRITE (unit_nr, *) "num_davidson_iter", num_davidson_iter
+         WRITE (unit_nr, *) "eps_exc_en", eps_exc_en
+         WRITE (unit_nr, *) "bse_davidson_abort_cond", bse_davidson_abort_cond
+         WRITE (unit_nr, *) "z_space_energy_cutoff", z_space_energy_cutoff
+      END IF
+
+      !MG to del
```
Comment (fstein93) [inline src/bse_iterative.F]: Please remove this debug code and similar code below

File: src/bse_iterative.F [hunk 2]
```diff
@@ -0,0 +1,1788 @@
...
+      IF (num_Z_vectors + num_new_t .LE. num_max_z_space) THEN
+         W_vectors(:, :, :num_Z_vectors) = Z_vectors(:, :, :)
+         W_vectors(:, :, num_Z_vectors + 1:) = Subspace_add_dir
+         num_Z_vectors = num_Z_vectors + num_new_t
+      ELSE
+         IF (unit_nr > 0) THEN
+            WRITE (unit_nr, *) "MG: Resetting dimension in i_iter=", i_iter
+         END IF
+         IF (unit_nr > 0) THEN
+            OPEN (10, file=res_filename, status="old", access="sequential", position="append")
```
Comment (fstein93) [inline src/bse_iterative.F]: Please ask CP2K for a unit number (close_file/open_file/get_unit_number from the module "cp_files") or use the provide ones (default_input_unit/default_output_unit from the module "machine")

File: src/bse_util.F [hunk 3]
```diff
@@ -0,0 +1,1350 @@
...
+                          col_indices=col_indices)
+
+      IF (unit_nr > 0) THEN
+         WRITE (unit_nr, *) header
+      END IF
+      IF (my_abs_vals) THEN
+         DO i = 1, nrow_local
+            DO j = 1, ncol_local
+               IF (ABS(fm%local_data(i, j)) > thresh) THEN
+                  WRITE (*, "(A7,T10,I5,T20,I5,T30,F13.5)") header, row_indices(i), col_indices(j), &
```
Comment (fstein93) [inline src/bse_util.F]: Add suitable unit numbers

File: src/input_cp2k_mp2.F [hunk 4]
```diff
@@ -852,15 +854,15 @@ SUBROUTINE create_ri_g0w0(section)
...
-      CALL keyword_create(keyword, __LOCATION__, name="BSE", &
-                          description="If true, electronic excitation energies are computed from the "// &
-                          "Bethe-Salpeter equation on top of GW eigenvalues.  Parameter of BSE can be adjusted in "// &
-                          "the corresponding section.", &
-                          usage="BSE", &
-                          default_l_val=.FALSE., &
-                          lone_keyword_l_val=.TRUE.)
-      CALL section_add_keyword(section, keyword)
-      CALL keyword_release(keyword)
+      ! CALL keyword_create(keyword, __LOCATION__, name="BSE", &
```
Comment (fstein93) [inline src/input_cp2k_mp2.F]: Why do you have a keyword and a section? Remove it if it is obsolete and not expected to be reused.

**Review by fstein93 — COMMENTED** [review]
I have left a few formal comments. The CI will complain about most of them anyways.

**Comment by MGraml** [comment]
Thank you for the review! I incorporated all your comments in the new commit.

**Comment by fstein93** [comment]
There is still the `write` statement from bse_util.F left.

**Pull request merged**